dashboard/app/models/concerns/curriculum/course_types.rb

```diff
@@ -15,6 +15,28 @@ module Curriculum::CourseTypes
     validates :participant_audience, acceptance: {accept: SharedCourseConstants::PARTICIPANT_AUDIENCE.to_h.values, message: 'must be facilitator, teacher, or student'}
 
     validate :cannot_have_same_audiences
+    validate :must_have_same_course_type_as_family
+  end
+
+  # All courses in the same family name must have the save instruction_type, instructor_audience, and participant audience
+  def must_have_same_course_type_as_family
+    all_family_courses = nil
+    family_name = self.family_name
+
+    if (is_a?(Script) && unit_group) || is_a?(UnitGroup)
+      family_name = unit_group.family_name if is_a?(Script) && unit_group
+      return if family_name.nil_or_empty?
+      all_family_courses = UnitGroup.all.select {|c| c.family_name == family_name}
+    elsif is_a?(Script)
+      return if family_name.nil_or_empty?
+      all_family_courses = Script.get_family_from_cache(family_name)
+    end
```

Comment:
sorry to pick on your code here, but I am having some trouble following this logic. it seems like there are several things happening across the same set of if-else conditions: computing family name, returning early, and computing all_family_courses. can you please try extracting a methods `get_family_name` and `get_family_courses`? I think that would help simplify things a lot.
```
family_name = get_family_name
return if family_name.nil_or_empty?
all_family_courses = get_family_courses(family_name)
```
Also I think this logic is complex enough that it deserves some unit tests. once you've extracted these helpers, it may make sense to write some of your tests against those helpers, and then not need to write so many tests against the outer method.

Reply:
Thanks for the suggestion on a way to improve the code. I will definitely do that.

Oooops on the tests. Totally meant to do that and then spaced... 🧠 is already on break I think